File: 0001~0500/0224. Basic Calculator/main.cpp
/** 
 * Source: https://leetcode.com/problems/basic-calculator/
 * Date: 2022/11/20
 * Skill: 
 * Runtime: 
 * Memory Usage: 
 * Time complexity: 
 * Space complexity: 
 * Constraints: 
 *      Every number and running calculation will fit in a signed 32-bit integer.
 *      There will be no two consecutive operators in the input.
 */

#include <iostream>
#include <stack>
#include <utility>
#include <string> 
using namespace std;

class Solution {
public:
    int calculate(string s) {
        int res = 0;
        stack<pair<char,int>> mystack;
        for (int i = 0; i < s.length(); i++) {

        }
    }

};

int main() {
    
}

static const auto io_sync_off = []() {
    // turn off sync
    std::ios::sync_with_stdio(false);
    // untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

File: 0001~0500/0224. Basic Calculator/main.py
"""
Source: https://leetcode.com/problems/basic-calculator/
Date: 2022/12/17
Skill: 
Runtime: 126 ms, faster than 74.67% 
Memory Usage: 15.5 MB, less than 33.58%
Time complexity: 
Space complexity: 
Constraints: 
    s consists of digits, '+', '-', '(', ')', and ' '.
"""

from typing import List

class Solution:
    def calculate(self, s: str) -> int:
        stack, nums, sign, sum, i = [], "0123456789", 1, 0, -1
        while i+1 < len(s):
            i += 1
            if s[i] == " ":
                continue
            elif s[i] in nums:
                num = s[i]
                while i+1 < len(s) and s[i+1] in nums:
                    i += 1
                    num += s[i]
                sum += sign*int(num)
                sign = 1
            elif s[i] == "(":
                stack.append((sum, sign))
                sum = 0
                sign = 1
            elif s[i] == ")":
                sum = stack[-1][0] + stack[-1][1]*sum
                stack.pop()
            elif s[i] == "-":
                sign *= -1
                
        return sum